includes/AMP/Traits/Sanitization_Utils.php

```diff
@@ -114,6 +114,139 @@ private function transform_a_tags( &$document ): void {
 		}
 	}
 
+	/**
+	 * Transforms all paragraphs in a story to use semantic heading tags if needed.
+	 *
+	 * This logic here mirrors the getTextElementTagNames() function in the editor
+	 * in order to change simple <p> tags into <h1>, <h2> or <h3>, depending on font size.
+	 *
+	 * Only relevant for older stories that haven't been updated in a while.
+	 *
+	 * @since 1.18.0
+	 *
+	 * @param Document|AMP_Document $document   Document instance.
+	 * @param bool                  $is_enabled Whether the feature is enabled.
+	 * @return void
+	 */
+	private function use_semantic_heading_tags( &$document, $is_enabled ): void {
+		if ( ! $is_enabled ) {
+			return;
+		}
+
+		$pages = $document->getElementsByTagName( 'amp-story-page' );
+
+		/**
+		 * The <amp-story-page> element
+		 *
+		 * @var DOMElement $page The <amp-story-page> element
+		 */
+		foreach ( $pages as $page ) {
+			$h1 = $page->getElementsByTagName( 'h1' );
+			$h2 = $page->getElementsByTagName( 'h2' );
+			$h3 = $page->getElementsByTagName( 'h3' );
+
+			// When a page already contains any headings, we don't need to do anything further.
+			if ( $h1->count() || $h2->count() || $h3->count() ) {
+				continue;
+			}
+
+			$text_elements = $document->xpath->query( './/p[ contains( @class, "text-wrapper" ) ]', $page );
+
+			if ( ! $text_elements ) {
+				return;
+			}
+
+			$this->use_semantic_heading_tags_for_elements( $text_elements );
+		}
+	}
+
+	/**
+	 * Transforms a list of elements to use semantic heading tags if needed.
+	 *
+	 * @since 1.18.0
+	 *
+	 * @param DOMNodeList $text_elements List of text elements.
+	 * @return void
+	 */
+	private function use_semantic_heading_tags_for_elements( $text_elements ): void {
+		// Matches PAGE_HEIGHT in the editor, as also seen in amp-story-grid-layer[aspect-ratio].
+		$page_height = 618;
+
+		$has_h1 = false;
+
+		/**
+		 * The individual text element.
+		 *
+		 * @var DOMElement $text_el The text element.
+		 */
+		foreach ( $text_elements as $text_el ) {
+			$style   = $text_el->getAttribute( 'style' );
+			$matches = [];
+
+			if ( ! preg_match( '/font-size:([^em]+)em/', $style, $matches ) ) {
+				continue;
+			}
+
+			// Contains the font-size in em.
+			// This is basically reversing the dataToFontSizeY() logic. Example:
+			// 0.582524em roughly equals 36 editor pixels: 0.582524 * 618 / 10 = 35.9999px.
+			$font_size_in_em = $matches[1];
+			$font_size_in_px = round( $font_size_in_em * $page_height / 10, 0 );
+
+			if ( $font_size_in_px >= 36 && ! $has_h1 ) {
+				$this->change_tag_name( $text_el, 'h1' );
+				$has_h1 = true;
+				continue;
+			}
+
+			if ( $font_size_in_px >= 27 ) {
+				$this->change_tag_name( $text_el, 'h2' );
+			} elseif ( $font_size_in_px >= 21 ) {
```

Comment:
Wondering if H3s should cover more sizes (21-26) vs H2s will take all the other sizes once there's an H1

When I tested I ended up with a single H3

<img width="516" alt="Screen Shot 2022-02-15 at 5 52 05 AM" src="https://user-images.githubusercontent.com/62242/154047604-6699a2d0-c483-4ef1-bf36-1ced2584534e.png">

Reply:
Great question. Since this is just a copy of the original implementation on the JS side to match our text presets (see the epic for context), any discussion about the algorithm is best held in a separate ticket.